lib/mongoid/clients.rb

```diff
@@ -61,7 +61,12 @@ def disconnect
       #
       # @since 3.0.0
       def with_name(name)
-        clients[name.to_sym] ||= Clients::Factory.create(name)
+        name_as_symbol = name.to_sym
+        return clients[name_as_symbol] if clients[name_as_symbol]
+        CREATE_LOCK.synchronize do
+          return clients[name_as_symbol] if clients[name_as_symbol]
+          clients[name_as_symbol] = Clients::Factory.create(name)
```

Comment:
Since you're inside the lock, you can use 
```ruby
CREATE_LOCK.synchronize do
  clients[name_as_symbol] ||= Clients::Factory.create(name)
end
```

as the `||=` is protected by it.